src/main/java/co/thatch/example/metrics/InMemoryMetricsService.java

```diff
@@ -0,0 +1,117 @@
+package co.thatch.example.metrics;
+
+import co.thatch.example.metrics.model.Metric;
+import co.thatch.example.metrics.model.MetricBucket;
+import io.micronaut.runtime.server.event.ServerStartupEvent;
+import io.micronaut.runtime.event.annotation.EventListener;
+import io.micronaut.scheduling.annotation.Async;
+
+import javax.inject.Singleton;
+import java.time.Duration;
+import java.time.Instant;
+import java.util.*;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.atomic.DoubleAdder;
+import java.util.stream.LongStream;
+
+
+@Singleton
+public class InMemoryMetricsService implements MetricsService {
+    private static final int MAX_WINDOW_SIZE = 1_800;
+    private static final int MAX_WINDOW_SIZE_WITH_BUFFER = MAX_WINDOW_SIZE + 1;
+
+    private final DoubleAdder requestCounter = new DoubleAdder();
+    private final double[] requestCountRecordings = new double[MAX_WINDOW_SIZE_WITH_BUFFER];
+    private final HashMap<Long, MetricHolder> metrics = new HashMap<>();
+    private final PriorityQueue<Long> timestamps = new PriorityQueue<>();
+
+    @Async
+    @EventListener
```

Comment:
This feels a little overkill for calculating the value in the current bucket? Is it possible to just update the value directly when recording the request?

Reply:
The scenario this is intended to cater for is for periods where no requests are received, resulting in stale entries in the recordings - creating incorrect QPS calculations